File: python/templates/outbound-ops/app/reads.py
from instantly.resources.account import list_account
from instantly.resources.campaign import get_campaign_analytics_overview
from instantly.resources.email import count_unread_emails


def items_of(res):
    """Extract a list from a paginated response without assuming the exact envelope."""
    if isinstance(res, list):
        return res
    if isinstance(res, dict):
        for key in ("items", "data", "accounts", "results"):
            if isinstance(res.get(key), list):
                return res[key]
    return None


def sender_health(client):
    """Read sender health. Negative status (connection/bounce error) or negative
    warmup_status (banned/spam/suspension) means the sender is in trouble
    (docs/api/accounts.md). Conservative: total=None if the payload is unfamiliar
    (never a false "0 healthy")."""
    res = list_account(client, {"query": {"limit": 100}})
    lst = items_of(res)
    if lst is None:
        return {"total": None, "green": 0, "flagged": [], "raw": res}
    flagged = []
    for a in lst:
        acc = a if isinstance(a, dict) else {}
        problems = []
        status = acc.get("status")
        warmup = acc.get("warmup_status")
        if isinstance(status, (int, float)) and status < 0:
            problems.append(f"status {status}")
        if isinstance(warmup, (int, float)) and warmup < 0:
            problems.append(f"warmup {warmup}")
        if problems:
            flagged.append({"email": acc.get("email", "(unknown)"), "reason": ", ".join(problems)})
    return {"total": len(lst), "green": len(lst) - len(flagged), "flagged": flagged}


def replies_waiting(client):
    res = count_unread_emails(client, {})
    if isinstance(res, (int, float)):
        return int(res)
    if isinstance(res, dict):
        for k in ("count", "unread"):
            if isinstance(res.get(k), (int, float)):
                return int(res[k])
    return None


def overview(client):
    return get_campaign_analytics_overview(client, {"query": {}})

File: js/templates/outbound-ops/lib/reads.ts
import {
  listAccount,
  getCampaignAnalyticsOverview,
  countUnreadEmails,
  type InstantlyClient,
} from "@instantly-ai/sdk"

/** Extract a list from a paginated response without assuming the exact envelope. */
export function itemsOf(res: unknown): unknown[] | null {
  if (Array.isArray(res)) return res
  const r = res as Record<string, unknown> | null
  for (const k of ["items", "data", "accounts", "results"]) {
    if (r && Array.isArray(r[k])) return r[k] as unknown[]
  }
  return null
}

export interface SenderHealth {
  total: number | null // null = couldn't parse the accounts payload
  green: number
  flagged: { email: string; status?: number; warmupStatus?: number; reason: string }[]
  raw?: unknown
}

/**
 * Read sender health from listAccount. Uses the documented codes (docs/api/accounts.md):
 * a negative `status` (connection/bounce/sending error) or negative `warmup_status`
 * (banned/spam/suspension) means the sender is in trouble. Conservative: only flags
 * what it can read; if the envelope is unfamiliar it returns total=null (never a false "0 healthy").
 */
export async function senderHealth(client: InstantlyClient): Promise<SenderHealth> {
  const res = await listAccount(client, { query: { limit: 100 } })
  const list = itemsOf(res)
  if (!list) return { total: null, green: 0, flagged: [], raw: res }
  const flagged: SenderHealth["flagged"] = []
  for (const a of list) {
    const acc = a as { email?: string; status?: number; warmup_status?: number }
    const problems: string[] = []
    if (typeof acc.status === "number" && acc.status < 0) problems.push(`status ${acc.status}`)
    if (typeof acc.warmup_status === "number" && acc.warmup_status < 0) problems.push(`warmup ${acc.warmup_status}`)
    if (problems.length) {
      flagged.push({ email: acc.email ?? "(unknown)", status: acc.status, warmupStatus: acc.warmup_status, reason: problems.join(", ") })
    }
  }
  return { total: list.length, green: list.length - flagged.length, flagged }
}

/** Count replies waiting in the inbox (defensive about the response shape). */
export async function repliesWaiting(client: InstantlyClient): Promise<number | null> {
  const res = await countUnreadEmails(client, {})
  if (typeof res === "number") return res
  const r = res as { count?: number; unread?: number } | null
  if (r && typeof r.count === "number") return r.count
  if (r && typeof r.unread === "number") return r.unread
  return null
}

/** Yesterday's campaign analytics overview (all campaigns). Shape varies — the renderer reads it defensively. */
export async function overview(client: InstantlyClient): Promise<unknown> {
  return getCampaignAnalyticsOverview(client, { query: {} })
}
